lib/ansible/modules/unarchive.py

```diff
@@ -952,15 +952,15 @@ def __init__(self, src, b_dest, file_args, module):
 # try handlers in order and return the one that works or bail if none work
 def pick_handler(src, dest, file_args, module):
     handlers = [ZipArchive, TgzArchive, TarArchive, TarBzipArchive, TarXzArchive, TarZstdArchive]
-    reasons = set()
+    reasons = []
     for handler in handlers:
         obj = handler(src, dest, file_args, module)
         (can_handle, reason) = obj.can_handle_archive()
         if can_handle:
             return obj
-        reasons.add(reason)
-    reason_msg = ' '.join(reasons)
-    module.fail_json(msg='Failed to find handler for "%s". Make sure the required command to extract the file is installed. %s' % (src, reason_msg))
+        reasons.append("%s: %s" % (handler.__name__, reason))
```

Comment:
The class names are arbitrary (only useful to those looking at the code) and can be determined from the executable (which should be in the reason), so I think this addition is overkill.

Reply:
no problem, I thought it would be clearer with the class name than with "/bin/tar" and a mention of the compression algorithm further down the line.